File: runtime/microvm/registry_client.go
package microvm

import (
	"context"
	"time"
)

const defaultSessionRegistryTTL = time.Hour

// RegistryClient is a deterministic constrained MicroVM client backed by a durable SessionRegistry.
type RegistryClient struct {
	registry SessionRegistry
	ttl      time.Duration
}

var _ Client = (*RegistryClient)(nil)

// RegistryClientOption configures a registry-backed constrained client.
type RegistryClientOption func(*RegistryClient)

// WithRegistryClientTTL configures the TTL applied to newly-created durable sessions.
func WithRegistryClientTTL(ttl time.Duration) RegistryClientOption {
	return func(client *RegistryClient) {
		if ttl > 0 {
			client.ttl = ttl
		}
	}
}

// NewRegistryClient creates a constrained client that persists sessions through a SessionRegistry.
func NewRegistryClient(registry SessionRegistry, opts ...RegistryClientOption) (*RegistryClient, error) {
	if registry == nil {
		return nil, safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm registry client requires a session registry", "")
	}
	client := &RegistryClient{registry: registry, ttl: defaultSessionRegistryTTL}
	for _, opt := range opts {
		if opt != nil {
			opt(client)
		}
	}
	return client, nil
}

// Create stores a requested durable session record.
func (c *RegistryClient) Create(ctx context.Context, input CreateSessionInput) (SessionRecord, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	if c == nil || c.registry == nil {
		return SessionRecord{}, safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm registry client requires a session registry", input.RequestID)
	}
	now := registryClientTime(input.Now)
	record := SessionRecord{
		TenantID:            input.TenantID,
		Namespace:           input.Namespace,
		SessionID:           input.SessionID,
		State:               StateRequested,
		DesiredState:        StateRequested,
		ProviderID:          DefaultSessionProviderID,
		ProviderMicroVMID:   input.SessionID,
		ProviderState:       string(StateRequested),
		AWSLifecycleState:   string(StateRequested),
		ImageRef:            input.ImageRef,
		NetworkConnectorRef: input.NetworkConnectorRef,
		ControllerID:        input.ControllerID,
		CreatedAt:           now,
		UpdatedAt:           now,
		LastObservedAt:      now,
		ExpiresAt:           now.Add(c.ttl),
		Generation:          1,
		LastAction:          CommandCreate,
		LastCommandID:       input.RequestID,
		AuthSubject:         input.AuthSubject,
		Metadata:            cloneStringMap(input.SessionSpec.Metadata),
	}
	return c.registry.Put(ctx, record)
}

// Start marks a durable session as starting toward started.
func (c *RegistryClient) Start(ctx context.Context, input SessionCommandInput) (SessionRecord, error) {
	return c.transition(ctx, input, CommandStart, StateStarting, input.DesiredState)
}

// Stop marks a durable session as stopping toward stopped.
func (c *RegistryClient) Stop(ctx context.Context, input SessionCommandInput) (SessionRecord, error) {
	return c.transition(ctx, input, CommandStop, StateStopping, input.DesiredState)
}

// Status returns durable status for a session.
func (c *RegistryClient) Status(ctx context.Context, input SessionQueryInput) (SessionStatus, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	record, err := c.session(ctx, input)
	if err != nil {
		return SessionStatus{}, err
	}
	status := SessionStatus{
		TenantID:        record.TenantID,
		Namespace:       record.Namespace,
		SessionID:       record.SessionID,
		State:           record.State,
		DesiredState:    record.DesiredState,
		LifecycleState:  record.State,
		Endpoint:        record.Endpoint,
		MicroVMID:       record.MicroVMID,
		LastAction:      record.LastAction,
		LastTransition:  record.UpdatedAt,
		RegistryVersion: record.Generation,
	}
	if err := ValidateSessionStatus(status); err != nil {
		return SessionStatus{}, err
	}
	return status, nil
}

// Session returns the durable session record.
func (c *RegistryClient) Session(ctx context.Context, input SessionQueryInput) (SessionRecord, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	return c.session(ctx, input)
}

func (c *RegistryClient) transition(ctx context.Context, input SessionCommandInput, action Command, state LifecycleState, desired LifecycleState) (SessionRecord, error) {
	if ctx == nil {
		ctx = context.Background()
	}
	if c == nil || c.registry == nil {
		return SessionRecord{}, safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm registry client requires a session registry", input.RequestID)
	}
	record, err := c.registry.Get(ctx, SessionKey{TenantID: input.TenantID, Namespace: input.Namespace, SessionID: input.SessionID})
	if err != nil {
		return SessionRecord{}, err
	}
	record.State = state
	record.DesiredState = desired
	record.ProviderID = defaultString(record.ProviderID, DefaultSessionProviderID)
	record.ProviderMicroVMID = defaultString(record.ProviderMicroVMID, record.SessionID)
	record.ProviderState = string(state)
	record.AWSLifecycleState = string(state)
	record.ControllerID = input.ControllerID
	record.AuthSubject = input.AuthSubject
	record.LastAction = action
	record.LastCommandID = input.RequestID
	record.UpdatedAt = registryClientTime(input.Now)
	record.LastObservedAt = record.UpdatedAt
	record.Generation++
	return c.registry.Put(ctx, record)
}

func (c *RegistryClient) session(ctx context.Context, input SessionQueryInput) (SessionRecord, error) {
	if c == nil || c.registry == nil {
		return SessionRecord{}, safeError(ErrorCodeSessionRegistryIncomplete, "apptheory: microvm registry client requires a session registry", input.RequestID)
	}
	return c.registry.Get(ctx, SessionKey{TenantID: input.TenantID, Namespace: input.Namespace, SessionID: input.SessionID})
}

func registryClientTime(value time.Time) time.Time {
	if !value.IsZero() {
		return value.UTC()
	}
	return time.Unix(0, 0).UTC()
}

func defaultString(value, fallback string) string {
	if value != "" {
		return value
	}
	return fallback
}

File: ts/src/microvm/registry-client.ts
import {
  MICROVM_DEFAULT_SESSION_PROVIDER_ID,
  MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
  MicroVMCommand,
  MicroVMState,
  type MicroVMClient,
  type MicroVMCommandName,
  type MicroVMLifecycleState,
  type MicroVMRegistryClientOptions,
  type MicroVMSessionCommandInput,
  type MicroVMSessionQueryInput,
  type MicroVMSessionRecord,
  type MicroVMSessionRegistry,
  type MicroVMSessionStatus,
  type MicroVMCreateSessionInput,
} from "./model.js";
import { safeError } from "./errors.js";
import { validateMicroVMSessionStatus } from "./session.js";
import { cloneStringMap } from "./safety.js";
import { coalesceMicroVMTime } from "./time.js";

export class MicroVMRegistryClient implements MicroVMClient {
  private readonly registry: MicroVMSessionRegistry;
  private readonly ttlMs: number;

  constructor(
    registry: MicroVMSessionRegistry,
    options: MicroVMRegistryClientOptions = {},
  ) {
    if (!registry) {
      throw safeError(
        MICROVM_ERROR_SESSION_REGISTRY_INCOMPLETE,
        "apptheory: microvm registry client requires a session registry",
        "",
      );
    }
    this.registry = registry;
    const ttlMs = Math.trunc(Number(options.ttl_ms) || 0);
    this.ttlMs = ttlMs > 0 ? ttlMs : 60 * 60 * 1000;
  }

  async create(
    input: MicroVMCreateSessionInput,
  ): Promise<MicroVMSessionRecord> {
    const now = coalesceMicroVMTime(input.now, new Date(0));
    const record: MicroVMSessionRecord = {
      tenant_id: input.tenant_id,
      namespace: input.namespace,
      session_id: input.session_id,
      state: MicroVMState.Requested,
      desired_state: MicroVMState.Requested,
      endpoint: "",
      microvm_id: "",
      provider_id: MICROVM_DEFAULT_SESSION_PROVIDER_ID,
      provider_microvm_id: input.session_id,
      provider_state: MicroVMState.Requested,
      aws_lifecycle_state: MicroVMState.Requested,
      image_ref: input.image_ref,
      network_connector_ref: input.network_connector_ref,
      controller_id: input.controller_id,
      created_at: now,
      updated_at: now,
      last_observed_at: now,
      expires_at: new Date(now.valueOf() + this.ttlMs),
      generation: 1,
      last_action: MicroVMCommand.Create,
      last_command_id: input.request_id,
      auth_subject: input.auth_subject,
    };
    const metadata = cloneStringMap(input.session_spec.metadata);
    if (metadata) record.metadata = metadata;
    return await this.registry.put(record);
  }

  async start(
    input: MicroVMSessionCommandInput,
  ): Promise<MicroVMSessionRecord> {
    return await this.transition(
      input,
      MicroVMCommand.Start,
      MicroVMState.Starting,
      MicroVMState.Started,
    );
  }

  async stop(input: MicroVMSessionCommandInput): Promise<MicroVMSessionRecord> {
    return await this.transition(
      input,
      MicroVMCommand.Stop,
      MicroVMState.Stopping,
      MicroVMState.Stopped,
    );
  }

  async status(input: MicroVMSessionQueryInput): Promise<MicroVMSessionStatus> {
    const record = await this.session(input);
    const status: MicroVMSessionStatus = {
      tenant_id: record.tenant_id,
      namespace: record.namespace,
      session_id: record.session_id,
      state: record.state,
      desired_state: record.desired_state,
      lifecycle_state: record.state,
      endpoint: record.endpoint ?? "",
      microvm_id: record.microvm_id ?? "",
      last_action: record.last_action,
      last_transition: record.updated_at,
      registry_version: record.generation,
    };
    validateMicroVMSessionStatus(status);
    return status;
  }

  async session(
    input: MicroVMSessionQueryInput,
  ): Promise<MicroVMSessionRecord> {
    return await this.registry.get({
      tenant_id: input.tenant_id,
      namespace: input.namespace,
      session_id: input.session_id,
    });
  }

  private async transition(
    input: MicroVMSessionCommandInput,
    action: MicroVMCommandName,
    state: MicroVMLifecycleState,
    desiredState: MicroVMLifecycleState,
  ): Promise<MicroVMSessionRecord> {
    const record = await this.registry.get({
      tenant_id: input.tenant_id,
      namespace: input.namespace,
      session_id: input.session_id,
    });
    const next: MicroVMSessionRecord = {
      ...record,
      state,
      desired_state: desiredState,
      provider_id: record.provider_id || MICROVM_DEFAULT_SESSION_PROVIDER_ID,
      provider_microvm_id: record.provider_microvm_id || record.session_id,
      provider_state: state,
      aws_lifecycle_state: state,
      controller_id: input.controller_id,
      auth_subject: input.auth_subject,
      last_action: action,
      last_command_id: input.request_id,
      updated_at: coalesceMicroVMTime(input.now, new Date(0)),
      last_observed_at: coalesceMicroVMTime(input.now, new Date(0)),
      generation: record.generation + 1,
    };
    return await this.registry.put(next);
  }
}

export function createMicroVMRegistryClient(
  registry: MicroVMSessionRegistry,
  options: MicroVMRegistryClientOptions = {},
): MicroVMRegistryClient {
  return new MicroVMRegistryClient(registry, options);
}
